### DISCUSSION TITLE
Formal specification

### DISCUSSION THREAD

**Discussion by sgiath** [markqvist/Reticulum#415] [General]
Hello, thanks for the fantastic project. I have been looking for something like this in [Meshtastic](https://meshtastic.org/), but I have always thought it could be pushed further.
My question is: is there some formal specification of the protocol? I have started implementing the Elixir port of the Reticulum, but I am missing a specification I could base it on. Right now, I work from Python code, but the last time I coded something in Python was at least five years ago, so it is not a very effective way of working. I know there is documentation; for example, the packet's wire format is very well documented, but other parts of the protocol are missing the specificity needed to fully implement it. For inspiration, I find the specification of the Noise protocol very approachable while being precise, and I was able to implement the protocol just using this spec: https://noiseprotocol.org/noise.html

**Comment by faragher** [comment]
Some of the information are defined in various classes, such as control
codes or the fields in an LXM, but some things, in my experience, you just
have to hunt down in the code. It would be helpful to have a formal
specification but as far as I'm aware, none yet exists.

**Reply by sgiath** [reply]
Thanks. Would you happen to know if something is planned in this area? I see in the roadmap that some overhaul to docs is planned, but I am not sure if it means formal specs or just user docs https://github.com/markqvist/Reticulum/blob/master/Roadmap.md#comprehensibility

Should I gather the information I will collect during the implementation and start putting it into the rough shape of formal specification? Is that something that would be useful or not?

**Comment by Erethon** [comment]
I think that we've reached a critical mass of people trying to implement the reticulum protocol in other language stacks (Rust, C++, Elixir now) that we'll soon have to start working on a proper spec. This is something I personally would like to start working on after 37C3.

**Comment by markqvist** [comment]
Hello @Sgiath! Thanks a bunch for looking into an Elixir port, that would be great! At this point, the Python reference implementation *is* the formal spec ;) At least, it's the best we have, and the only way I could realistically make this work within a reasonable amount of time.

Writing a full natural-language formal spec would definitely be very, very helpful for everyone at this point.
